eknrenderer/eknrenderer/eknr-renderer.c

```diff
@@ -0,0 +1,1312 @@
+ * @file: A #GFile specifying the location of the template file
+ * @variables: The variables and sections to use when rendering.
+ * @error: A #GError
+ *
+ * Use mustache_c to render a document, similar to
+ * eknr_renderer_render_mustache_document, but read the template
+ * from the file specified at @file. If that file has already been
+ * read, its contents will be read from the internal cache.
+ *
+ * Returns: The renderered document on success, %NULL on error.
+ */
+char *
+eknr_renderer_render_mustache_document_from_file (EknrRenderer *renderer,
+                                                  GFile        *file,
+                                                  GVariant     *variables,
+                                                  GError      **error)
+{
+  g_autofree char *uri = g_file_get_uri (file);
+  EknrRendererPrivate *priv = eknr_renderer_get_instance_private (renderer);
+  mustache_template_t *tmpl = g_hash_table_lookup (priv->cache, uri);
+  g_autofree char *contents = NULL;
+  g_autoptr(RendererMustacheData) data = NULL;
+
+  if (tmpl != NULL)
+    return _renderer_render_mustache_document_internal (tmpl, variables, error);
+
+  if (!g_file_load_contents (file, NULL, &contents, NULL, NULL, error))
+    return NULL;
+
+  data = renderer_mustache_data_new (NULL,
+                                     error,
+                                     contents);
+  tmpl = mustache_compile (&_renderer_mustache_data_vfuncs, (char *) data);
+
+  if (tmpl == NULL)
+    return NULL;
+
+  g_hash_table_replace (priv->cache, g_steal_pointer (&uri), tmpl);
+
+  return _renderer_render_mustache_document_internal (tmpl, variables, error);
+}
+
+/**
+ * eknr_renderer_render_mustache_document:
+ * @renderer: An #EknrRenderer
+ * @tmpl_text: The template to render
+ * @variables: The variables and sections to use when rendering.
+ * @error: A #GError
+ *
+ * Use mustache_c to render a document. The provided @variables variant
+ * is used as substitutions. The variant should be of type
+ * 'a{sv}' and each child node should be either 's' or a variable subsitution
+ * 'as' for a section substitution.
+ *
+ * Returns: The renderered document on success, %NULL on error.
+ */
+char *
+eknr_renderer_render_mustache_document (EknrRenderer *renderer,
+                                        const char   *tmpl_text,
+                                        GVariant     *variables,
+                                        GError      **error)
+{
+  g_autoptr(GVariantDict) variables_dict = g_variant_dict_new (variables);
+  g_autoptr(RendererMustacheData) data = renderer_mustache_data_new (variables_dict,
+                                                                     error,
+                                                                     tmpl_text);
+  g_autoptr(mustache_template_t) tmpl = mustache_compile (&_renderer_mustache_data_vfuncs,
+                                                          (char *) data);
```

Comment:
Pretty sure that this cast is wrong; either you need to cast to `(void *)` or if that's the case, most likely omit the cast altogether since it wouldn't be necessary.

Reply:
whoops, it is the wrong cast! I must have thought that `data` was a string when I was writing this, not a closure! I'll omit the cast.